src/dx/subscribe/subscription.rs

```diff
@@ -0,0 +1,751 @@
+//! # Subscription module.
+//!
+//! This module contains the [`Subscription`] type, which is used to manage
+//! subscription to the specific entity and attach listeners to process
+//! real-time events triggered for the `entity`.
+
+use spin::RwLock;
+use uuid::Uuid;
+
+use crate::core::{Deserializer, Transport};
+use crate::{
+    core::{DataStream, PubNubEntity},
+    dx::pubnub_client::PubNubClientInstance,
+    lib::{
+        alloc::{
+            string::String,
+            sync::{Arc, Weak},
+            vec,
+            vec::Vec,
+        },
+        collections::HashMap,
+        core::{
+            cmp::PartialEq,
+            fmt::{Debug, Formatter, Result},
+            ops::{Deref, DerefMut},
+        },
+    },
+    subscribe::{
+        event_engine::SubscriptionInput, traits::EventHandler, AppContext, EventDispatcher,
+        EventEmitter, EventSubscriber, File, Message, MessageAction, Presence, SubscribableType,
+        SubscriptionCursor, SubscriptionOptions, SubscriptionSet, Update,
+    },
+};
+
+/// Entity subscription.
+///
+/// # Example
+///
+/// ### Multiplexed subscription
+///
+/// ```rust
+/// use pubnub::{
+///     subscribe::{Subscriber, SubscriptionOptions}, Keyset, PubNubClient, PubNubClientBuilder,
+/// };
+///
+/// # fn main() -> Result<(), pubnub::core::PubNubError> {
+/// let client = // PubNubClient
+/// #     PubNubClientBuilder::with_reqwest_transport()
+/// #         .with_keyset(Keyset {
+/// #              subscribe_key: "demo",
+/// #              publish_key: Some("demo"),
+/// #              secret_key: Some("demo")
+/// #          })
+/// #         .with_user_id("uuid")
+/// #         .build()?;
+/// let channel = client.create_channel("my_channel");
+/// let subscription = channel.subscription(None);
+/// // Subscription with presence announcements
+/// let subscription_with_presence = channel.subscription(Some(vec![SubscriptionOptions::ReceivePresenceEvents]));
+/// #     Ok(())
+/// # }
+/// ```
+///
+/// ### Sum of subscriptions
+///
+/// ```rust
+/// use pubnub::{
+///     subscribe::{Subscriber, Subscription, SubscriptionSet},
+///     Keyset, PubNubClient, PubNubClientBuilder,
+/// };
+///
+/// # fn main() -> Result<(), pubnub::core::PubNubError> {
+/// let client = // PubNubClient
+/// #     PubNubClientBuilder::with_reqwest_transport()
+/// #         .with_keyset(Keyset {
+/// #              subscribe_key: "demo",
+/// #              publish_key: Some("demo"),
+/// #              secret_key: Some("demo")
+/// #          })
+/// #         .with_user_id("uuid")
+/// #         .build()?;
+/// let channels = client.create_channels(&["my_channel_1", "my_channel_2"]);
+/// let subscription = channels[0].subscription(None).add(channels[1].subscription(None));
```

Comment:
This chain is also very missleading. Can you explain what it exactly does?

Reply:
I added those arithmetic operations before realizing that I'll probably need to work with `Arc<Subscription>` and `Arc<SubscriptionSet>`. Creating a separate struct with a single value just to add it - it will cause naming hell or not? 
I was planning to remove those traits implementation.

Reply:
I mean, I totally don't understand what means `channels[0].subscription(None).add(channels[1].subscription(None))`
It seems like an old C-style code that contains some macro values like
`do_something(value, NULL, NULL, NULL); `
I can't tell what should be better in this case, but for random reader (e.g. @techwritermat ) it can be a little bit hard to understand what is happening here.

Reply:
I've made proper implementation, now there is simple additions :)